coalib/core/Core.py

```diff
@@ -0,0 +1,167 @@
+                running_tasks,
+                event_loop,
+                executor,
+                task):
+    """
+    The callback for when a task of a bear completes. It is responsible for
+    checking if the bear completed its execution and the handling of the
+    result generated by the task.
+
+    :param bear:
+        The bear that the task belongs to.
+    :param result_callback:
+        A callback function which is called when results are available.
+    :param running_tasks:
+        Dictionary that keeps track of the remaining tasks of each bear.
+    :param event_loop:
+        The ``asyncio`` event loop bear-tasks are scheduled on.
+    :param executor:
+        The executor to which the bear tasks are scheduled.
+    :param task:
+        The task that completed.
+    """
+    try:
+        results = task.result()
+    except Exception as ex:
+        # FIXME Try to display only the relevant traceback of the bear if error
+        # FIXME occurred there, not the complete event-loop traceback.
+        logging.error('An exception was thrown during bear execution.',
+                      exc_info=ex)
+
+        results = None
+    finally:
+        running_tasks[bear].remove(task)
+        cleanup_bear(bear, running_tasks, event_loop)
+
+    if results is not None:
+        for result in results:
+            try:
+                # FIXME Long operations on the result-callback could block the
+                # FIXME   scheduler significantly. It should be possible to
+                # FIXME   schedule new Python Threads on the given event_loop
+                # FIXME   and process the callback there.
+                result_callback(result)
+            except Exception as ex:
+                # FIXME Try to display only the relevant traceback of the
+                # FIXME result handler if error occurred there, not the
+                # FIXME complete event-loop traceback.
+                logging.error(
+                    'An exception was thrown during result-handling.',
+                    exc_info=ex)
+
+
+def run(bears, result_callback):
+    """
+    Runs a coala session.
+
+    :param bears:
+        The bear instances to run.
+    :param result_callback:
+        A callback function which is called when results are available. Must
+        have following signature::
+
+            def result_callback(result):
+                pass
+    """
+    # FIXME Allow to pass different executors nicely, for example to execute
+    # FIXME   coala with less cores, or to schedule jobs on distributed systems
+    # FIXME   (for example Mesos).
+
+    # Set up event loop and executor.
+    event_loop = asyncio.SelectorEventLoop()
+    executor = concurrent.futures.ProcessPoolExecutor(
+        max_workers=_get_cpu_count())
+
+    # Let's go.
+    schedule_bears(bears, result_callback, event_loop, {}, executor)
+    try:
+        event_loop.run_forever()
+    finally:
+        event_loop.close()
```

Comment:
If a bear exits or raises an error, other bears still continue to run. Happens here: https://github.com/coala/coala/pull/3871/commits/bb19717550d61898d2c6703354c5ca32c10fe5ba#diff-94f6470cc4dcfa1e33ff5eb050fd5191R112